sync2jira/upstream_issue.py

```diff
@@ -368,25 +378,38 @@ def _get_current_project_node(upstream, project_number, issue_number, gh_issue):
     project_items = gh_issue['projectItems']['nodes']
     # If there are no project items, there is nothing to return.
     if len(project_items) == 0:
-        log.debug("Issue %s/%s is not associated with any project", upstream, issue_number)
+        log.debug("Issue %s#%s is not associated with any project",
+                  upstream, issue_number)
         return None
-    # If there is exactly one project item, return it if we don't have a configured project.
-    if not project_number and len(project_items) == 1:
+    # If there is exactly one project item, return it.  Even if it doesn't
+    # match the configured project, having one is better than having none.
```

Comment:
If the user has explicitly specified a project, it is better to stick to that project for syncing. And if an issue is linked to some other project, that project may not be having the required fields, and the expected format.

Reply:
If `sync2jira` has been requested to sync a given issue (either via a fedmsg notification or because we're initializing a repo) and the issue is not associated with the project specified in the configuration, we have three options:

1. ignore the issue altogether
2. sync only the non-project fields of the issue
3. try to sync all the fields and ignore only the individual fields which don't "work"

The existing code uses the first option, and that resulted in a "user surprise" for me and Matt.  It seems like the tool should follow one of the latter two options.

I realize that sync'ing priorities might be problematic, because the upstream values are project-specific keywords and the downstream values are project-schema-dependent, but I believe the code is already equipped to handle the case where the configuration doesn't match the source or destination, so I figured it wasn't really a problem to try to sync the priorities -- at worst the downstream simply wouldn't update that field.

As for story points, they are just numbers, so, as long as we can source the value, we can update it.  However, the fact that it comes from the "wrong" project is arguably a good reason against trying to sync it, but I concluded that bad data was better than no data.  (There is a problem with visibility:  how does a developer lookin in Jira know whether it's in sync with the upstream?...I was hoping that bad data would be more obvious and motivating than no data...but, perhaps that's wrong.)

In any case, I think we should move away from option 1, but I'm happy to go to option 2 instead of option 3.  Thoughts?

Reply:
I think lack of data is more noticeable than wrong data, especially for something that we're hoping is fairly hands off for teams once they get things set up, but I'm happy with anything that isn't option 1 :) It would be easier to see at the aggregate level this data will be used for that there is data missing from synced issues, than to recognize the nuance of incorrectly synced data that pulled from an unexpected source. As you say though, there's no guarantee that the values would match between projects, so if we handle that gracefully, there may not be much difference between the two approaches after the issue is synced using any data that maps correctly.

I've seen data from our teams that shows not all issues in a repo are always included in the related project, so ignoring those issues altogether was a problem. Since sync2jira is mapped at the repository level and not a project, it makes sense to me that we'd sync using the base data we have available. I haven't seen off-hand that an issue is part of more than 1 project, but it could be possible. It could also be possible to have 1 project association for an issue, but have it not match the configured project number, although I don't remember seeing an example of this yet.

https://github.com/release-engineering/Sync2Jira/issues/171 could potentially help us identify ambiguous issues in the future, if someone decides to pick it up, by adding some mechanism that indicates some further triage may be needed.

Reply:
> there may not be much difference between the two approaches after the issue is synced using any data that maps correctly.

Except for the fact that it (i.e., storypoints) would be coming from the wrong source and therefore would quite possibly be wrong....

> I haven't seen off-hand that an issue is part of more than 1 project, but it could be possible. It could also be possible to have 1 project association for an issue, but have it not match the configured project number, although I don't remember seeing an example of this yet.

If you're really interested, I should be able to dig up examples of both.  There is at least one instance of the second in the `intellij-common` repo configuration that you and I were testing with.  And, I've hit instances of the first in the `eclipse-che` organization.

> https://github.com/release-engineering/Sync2Jira/issues/171 could potentially help us

Yes:  it would be great if there were some way to make problems visible (e.g., via Jira) to the end-users, who don't have ready access to the `sync2jira` logs.

I'll modify the code to use Option 2.